tiledb/sm/filesystem/s3.cc

```diff
@@ -262,43 +263,55 @@ Status S3::disconnect() {
+  multipart_upload_mtx_.read_lock();
 
   if (multipart_upload_states_.size() > 0) {
-    RETURN_NOT_OK(init_client());
-
-    for (auto& kv : multipart_upload_states_) {
-      const MultiPartUploadState& state = kv.second;
-
-      if (state.st.ok()) {
-        Aws::S3::Model::CompleteMultipartUploadRequest complete_request =
-            make_multipart_complete_request(state);
-        auto outcome = client_->CompleteMultipartUpload(complete_request);
-        if (!outcome.IsSuccess()) {
-          const Status st = LOG_STATUS(Status::S3Error(
-              std::string("Failed to disconnect and flush S3 objects. ") +
-              outcome_error_message(outcome)));
-          if (!st.ok()) {
-            ret_st = st;
+    RETURN_NOT_OK_ELSE(init_client(), multipart_upload_mtx_.read_unlock());
+
+    std::vector<const MultiPartUploadState*> states;
+    states.reserve(multipart_upload_states_.size());
+    for (auto& kv : multipart_upload_states_)
+      states.emplace_back(&kv.second);
+
+    auto statuses =
+        parallel_for(vfs_thread_pool_, 0, states.size(), [&](uint64_t i) {
+          const MultiPartUploadState* state = states[i];
+          // Lock multipart state
+          std::unique_lock<std::mutex> state_lck(state->mtx);
+
+          if (state->st.ok()) {
+            Aws::S3::Model::CompleteMultipartUploadRequest complete_request =
+                make_multipart_complete_request(*state);
+            auto outcome = client_->CompleteMultipartUpload(complete_request);
+            if (!outcome.IsSuccess()) {
+              const Status st = LOG_STATUS(Status::S3Error(
+                  std::string("Failed to disconnect and flush S3 objects. ") +
+                  outcome_error_message(outcome)));
+              if (!st.ok()) {
+                ret_st = st;
+              }
+            }
+          } else {
+            Aws::S3::Model::AbortMultipartUploadRequest abort_request =
+                make_multipart_abort_request(*state);
+            auto outcome = client_->AbortMultipartUpload(abort_request);
+            if (!outcome.IsSuccess()) {
+              const Status st = LOG_STATUS(Status::S3Error(
+                  std::string("Failed to disconnect and flush S3 objects. ") +
+                  outcome_error_message(outcome)));
+              if (!st.ok()) {
+                ret_st = st;
+              }
+            }
           }
-        }
-      } else {
-        Aws::S3::Model::AbortMultipartUploadRequest abort_request =
-            make_multipart_abort_request(state);
-        auto outcome = client_->AbortMultipartUpload(abort_request);
-        if (!outcome.IsSuccess()) {
-          const Status st = LOG_STATUS(Status::S3Error(
-              std::string("Failed to disconnect and flush S3 objects. ") +
-              outcome_error_message(outcome)));
-          if (!st.ok()) {
-            ret_st = st;
-          }
-        }
-      }
-    }
-  }
+          return Status::Ok();
```

Comment:
unlock `multipart_upload_mtx_`

Reply:
This is inside the parallel for so I don't think we want to unlock it here. We unlock it a few lines below after we check the statuses to know all state operations have been completed.

Reply:
you're right -- i didn't pay attention to where this return was. resolving